File: dp/0115-distinct-subsequences.go
package dp

// https://leetcode.com/problems/distinct-subsequences/

// Approach #3: Bottom-Up Dynamic Programming (Space Optimized)
// Time: O(ns * nt), ns=len(s), nt=len(t)
// Space: O(nt)
func numDistinct(s string, t string) int {
	var ns, nt = len(s), len(t)
	var dp = make([]int, nt+1)
	dp[0] = 1
	for is := 1; is <= ns; is++ {
		for it := nt; it > 0; it-- {
			if s[is-1] == t[it-1] {
				dp[it] += dp[it-1]
			}
		}
	}
	return dp[nt]
}

// // Approach #2: Bottom-Up Dynamic Programming
// // Time: O(ns * nt), ns=len(s), nt=len(t)
// // Space: O(ns * nt)
// func numDistinct(s string, t string) int {
// 	var ns, nt = len(s), len(t)
// 	var dp = make([][]int, ns+1)
// 	for idx := range dp {
// 		dp[idx] = make([]int, nt+1)
// 		dp[idx][0] = 1
// 	}
// 	for is := 1; is <= ns; is++ {
// 		for it := 1; it <= nt; it++ {
// 			if s[is-1] == t[it-1] {
// 				dp[is][it] = dp[is-1][it] + dp[is-1][it-1]
// 			} else {
// 				dp[is][it] = dp[is-1][it]
// 			}
// 		}
// 	}
// 	return dp[ns][nt]
// }

// // Approach #1: Top-Down Dynamic Programming
// // Time: O(ns * nt), ns=len(s), nt=len(t)
// // Space: O(ns * nt)
// func numDistinct(s string, t string) int {
// 	var ns, nt = len(s), len(t)
// 	var dp = make([][]*int, ns)
// 	for idx := range dp {
// 		dp[idx] = make([]*int, nt)
// 	}
// 	var dfs func(is int, it int) int
// 	dfs = func(is int, it int) int {
// 		if it == nt {
// 			return 1
// 		}
// 		if is == ns {
// 			return 0
// 		}
// 		if dp[is][it] != nil {
// 			return *dp[is][it]
// 		}
// 		dp[is][it] = new(int)
// 		if s[is] != t[it] {
// 			*dp[is][it] = dfs(is+1, it)
// 		} else {
// 			*dp[is][it] = dfs(is+1, it) + dfs(is+1, it+1)
// 		}
// 		return *dp[is][it]
// 	}
// 	return dfs(0, 0)
// }

File: dp/0115-distinct-subsequences.py
import unittest

# https://leetcode.com/problems/distinct-subsequences/

# python3 -m unittest dp/0115-distinct-subsequences.py

class Solution(unittest.TestCase):
    # # Approach: Top-Down Dynamic Programming
    # # Time: O(ns * nt), ns=len(s), nt=len(t)
    # # Space: O(ns * nt)
    # def numDistinct(self, s: str, t: str) -> int:
    #     ns, nt = len(s), len(t)
    #     dp = [[None] * nt for _ in range(ns)]
    #     def dfs(i: int, j: int) -> int:
    #         if j == nt: return 1
    #         if i == ns: return 0
    #         if dp[i][j] != None:
    #             return dp[i][j]
    #         if s[i] == t[j]:
    #             dp[i][j] = dfs(i+1, j) + dfs(i+1, j+1)
    #         else:
    #             dp[i][j] = dfs(i+1, j)
    #         return dp[i][j]
    #     return dfs(0, 0)

    # # Approach: Bottom-UP Dynamic Programming
    # # Time: O(ns * nt), ns=len(s), nt=len(t)
    # # Space: O(ns * nt)
    # def numDistinct(self, s: str, t: str) -> int:
    #     ns, nt = len(s), len(t)
    #     dp = [[0] * (nt+1) for _ in range(ns+1)]
    #     for i in range(ns+1): dp[i][0] = 1
    #     for i in range(1, ns+1):
    #         for j in range(1, nt+1):
    #             if s[i-1] == t[j-1]:
    #                 dp[i][j] = dp[i-1][j] + dp[i-1][j-1]
    #             else:
    #                 dp[i][j] = dp[i-1][j]
    #     return dp[ns][nt]

    # Approach: Bottom-UP Dynamic Programming (Space Optimized)
    # Time: O(ns * nt), ns=len(s), nt=len(t)
    # Space: O(nt)
    def numDistinct(self, s: str, t: str) -> int:
        ns, nt = len(s), len(t)
        dp = [0] * (nt+1)
        dp[0] = 1
        for i in range(1, ns+1):
            for j in range(nt, 0, -1):
                if s[i-1] == t[j-1]:
                    dp[j] += dp[j-1]
        return dp[nt]

    def test(self):
        for s, t, expected in [
            ("rabbbit", "rabbit", 3),
            ("babgbag", "bag", 5),
            ("a", "", 1),
            ("", "a", 0),
        ]:
            output = self.numDistinct(s, t)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
